Extension/src/LanguageServer/Providers/inlayHintProvider.ts

```diff
@@ -0,0 +1,144 @@
+        if (!inlayHintsResult.canceled) {
+            if (inlayHintsResult.fileVersion === openFileVersions.get(uriString)) {
+                const cacheEntry: InlayHintsCacheEntry = this.createCacheEntry(inlayHintsResult);
+                this.cache.set(uriString, cacheEntry);
+                return this.getHints(cacheEntry);
+            } else {
+                // Force another request because file versions do not match.
+                this.onDidChangeInlayHintsEvent.fire();
+            }
+        }
+        return undefined;
+    }
+
+    public invalidateFile(uri: string): void {
+        this.cache.delete(uri);
+        this.onDidChangeInlayHintsEvent.fire();
+    }
+
+    private getHints(cacheEntry: InlayHintsCacheEntry): vscode.InlayHint[] {
+        let result: vscode.InlayHint[] = [];
+        const showTypeHintsEnabled: boolean = true; // TODO: get value from settings.
+        const showParamHintsEnabled: boolean = true; // TODO: get value from settings.
+        if (showTypeHintsEnabled) {
+            result = result.concat(cacheEntry?.TypeHints);
+        }
+        if (showParamHintsEnabled) {
+            const resolvedParameterHints: vscode.InlayHint[] = this.resolveParameterHints(cacheEntry.ParameterHints);
+            result = result.concat(resolvedParameterHints);
+        }
+        return result;
+    }
+
+    private resolveParameterHints(hints: CppInlayHint[]): vscode.InlayHint[] {
+        const resolvedHints: vscode.InlayHint[] = [];
+        const showRefEnabled: boolean = true; // TODO: get from settings
+        const hideParamNameEnabled: boolean = false; // TODO: get from settings
+        hints.forEach((h: CppInlayHint) => {
+            // Build parameter label based on settings.
+            // TODO: remove label if param includes parameter name or in comments.
+            const paramName: string = (hideParamNameEnabled /* && h.hasParamName*/) ? "" : h.label;
+            let refString: string = "";
+            if (showRefEnabled && h.isValueRef) {
+                refString = (paramName.length > 0) ? "& " : "&";
+            }
+            const colonString: string = (paramName.length > 0 || refString.length > 0) ? ":" : "";
+            const label: string = refString + paramName + colonString;
+
+            const inlayHint: vscode.InlayHint = new vscode.InlayHint(
+                new vscode.Position(h.position.line, h.position.character),
+                label,
+                vscode.InlayHintKind.Parameter);
+            inlayHint.paddingRight = true;
+            resolvedHints.push(inlayHint);
+        });
+        return resolvedHints;
+    }
+
+    private createCacheEntry(inlayHintsResults: GetInlayHintsResult): InlayHintsCacheEntry {
+        const typeHints: vscode.InlayHint[] = [];
+        inlayHintsResults.inlayHints.forEach((h: CppInlayHint) => {

```

Comment:
Would a for loop be better? I think forEach is not recommended to use, although we some of our existing code does that still.

Reply:
Changed to `for...of` loop.